src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OSPSuite.Core.Domain.Services
+{
+   public interface IConcurrencyManager
+   {
+      /// <summary>
+      /// 
+      /// </summary>
+      /// <typeparam name="TData">Data type to consume by the worker function</typeparam>
+      /// <typeparam name="TResult">Data produced by the worker function</typeparam>
+      /// <param name="numberOfCoresToUse">Number of cores to use. Use 0 or negative to take all cores</param>
+      /// <param name="cancellationToken">Cancellation token to cancel the threads</param>
+      /// <param name="data">List of data to consume by the workers</param>
+      /// <param name="action">A function to run on each worker on each piece of data</param>
+      /// <returns>Dictionary binding a result for each input data after running the action on it</returns>
+      Task<Dictionary<TData, TResult>> RunAsync<TData, TResult>(int numberOfCoresToUse, CancellationToken cancellationToken, ConcurrentQueue<TData> data, Func<int, CancellationToken, TData, Task<TResult>> action);
+   }
+
+   public class ConcurrencyManager : IConcurrencyManager
+   {
+      private readonly ICoreUserSettings _coreUserSettings;
+
+      public ConcurrencyManager(ICoreUserSettings coreUserSettings)
+      {
+         _coreUserSettings = coreUserSettings;
+      }
+
+      public async Task<Dictionary<TData, TResult>> RunAsync<TData, TResult>(int numberOfCoresToUse, CancellationToken cancellationToken, ConcurrentQueue<TData> data, Func<int, CancellationToken, TData, Task<TResult>> action)
+      {
+         if (numberOfCoresToUse <= 0)
+            numberOfCoresToUse = _coreUserSettings.MaximumNumberOfCoresToUse;
+         numberOfCoresToUse = Math.Min(numberOfCoresToUse, data.Count);
+
+         var results = new ConcurrentDictionary<TData, TResult>();
+         //Starts one task per core
+         var tasks = Enumerable.Range(0, numberOfCoresToUse).Select(coreIndex =>
+            Task.Run(async () =>
+            {
+               //While there is data left
+               while (data.TryDequeue(out var datum))
+               {
+                  //Invoke the action on it and store the result
+                  var result = await action.Invoke(coreIndex, cancellationToken, datum);
+                  results.TryAdd(datum, result);
+               }
+            }, cancellationToken)
+         ).ToList();
+
+         await Task.WhenAll(tasks);
+         //all tasks are completed. Can return results
+
+         return results.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
```

Comment:
why not return results? Just curious

Reply:
So we return a Task you could await. Again, from R you cannot await so the ConcurrentSimulationRunner does not use this feature but from other parts of the code you could have async/await

Reply:
This is not what I mean.results is already a dictionary. Why are your returning another dictionary based of this? The async/await is created for you automatically because you are using await /async in your code.

Reply:
To the best of my knowledge, ConcurrentDictionary and Dictionary are different classes. They do not share a common interface or anything like that. We need a ConcurrentDictionary to avoid concurrent access issues but beyond ConcurrencyManager class, we do not need it anymore so we should turn it into a Dictionary. Similarly, we use a ConcurrentQueue. I have just updated the signature so it accepts a IEnumerable instead of a ConcurrentQueue since this data structure is only needed as a robust implementation but should not be affecting the signature of the method.

Reply:
>To the best of my knowledge, ConcurrentDictionary and Dictionary are different classes. They do not share a common interface or anything like that.

oh yeah they share IDictionary or IReadOnlyDictionary. 

IReadOnlyDictionary is what we should return in that case anyways so that the caller does not mutate (at least not without explicit casting) the results

![image](https://user-images.githubusercontent.com/1041237/115238753-07e36080-a0ec-11eb-9d24-8bc18214285e.png)